### ISSUE TITLE
Automatic re-establishing WebSocket connection won’t re-establish subscriptions

### ISSUE THREAD

**Issue by palant** [neet/masto.js#1237]
I’ve managed to reduce this issue to the following code (this is the Elk test account under https://universeodon.com/@elkdev/):

```js
import {createStreamingAPIClient} from 'https://cdn.jsdelivr.net/npm/masto@6.10.0/+esm';

const client = createStreamingAPIClient({
  streamingApiUrl: "wss://universeodon.com",
  accessToken: "yZcpj0FmnsEkUvBiXSCb_KQnccl2IU0kx9TfDbcxPJY",
  implementation: WebSocket
});

const stream = client.user.notification.subscribe();
setTimeout(() => stream.connector.ws.close(), 10000);
for await (let update of stream)
  console.log(update);
```

Here I am closing the WebSocket connection explicitly after 10 seconds, but it could be closed for any number of reasons. In principle, `masto` considers that scenario by automatically re-establishing the connection. However, it won’t re-subscribe on the new connection, so the loop will continue but won’t receive any more updates. For example, some account favoriting https://universeodon.com/@elkdev/109695898772972577 will result in a message being logged if it happens within the first 10 seconds, not after the connection is restored however.

Elk issue in question is https://github.com/elk-zone/elk/issues/3017.

**Comment by neet** [comment]
Hi @palant, thank you for your inspection.

I doubt that your repro code which calls `#close` explicitly to simulate network error does not seem to be related to the original issue, as `#close` is a method to dispose of the streaming API client without reconnection.

Also, you mentioned that Masto.js won't re-send a "subscribe" event after reconnection, but it does:
https://github.com/neet/masto.js/blob/23999e883e82536fc51fc623deb425327a583d93/src/adapters/ws/web-socket-subscription.ts#L29-L34

Nonetheless, your issue in elk-zone/elk#3017 is worth researching. Apparently, browsers don't seem to dispatch a "closed" event for network disconnection, but I'm unsure if there's another way to detect it.

**Comment by neet** [comment]
Hi @palant, thank you for filing an issue. It seems to be a regression caused by #1224; though I doubt your repro code which calls `close()` explicitly doesn't seem to be related to the original issue, because `close()` is something you can dispose of a client without any reconnection.

Let me take some time to look into this issue.

**Comment by palant** [comment]
Given that Elk is still on version 6.7.5, it’s definitely not a regression from this issue.

I did verify that with the code above masto (both v6.7.5 and v6.10.0) opens a new WebSocket connection which no longer receives any messages due to not being subscribed to anything. I called `WebSocket.close()` here because it’s easier to reproduce than waiting for the connection to be closed due to inactivity, but I’ll test the fix in both scenarios.

**Comment by neet** [comment]
I'm unsure when this bug started, but I closed it after verifying that the issue you described is no longer occurring. You can see the test for it here:
https://github.com/neet/masto.js/blob/d62d4c5ca665af686cf5c368a2bf2b9989a1cc91/src/adapters/ws/web-socket-connector.spec.ts#L45

I did not verify if this patch resolves elk-zone/elk#3017, so please feel free to reopen this issue if you continue to experience problems.
